### PR TITLE
Issue 4710: physics.raycast now can return all results

### LINKED ISSUE #4710: Add option to let ray casts return all objects along the ray
https://forum.defold.com/t/physics-raycast-must-return-all-objects-along-the-ray/62933/8?u=britzl

### PR DESCRIPTION
**Pull request by JCash**

### CODE REVIEW

File: engine/gamesys/src/gamesys/components/comp_collision_object.cpp [hunk 1]
```diff
@@ -1,10 +1,10 @@
 // Copyright 2020 The Defold Foundation
 // Licensed under the Defold License version 1.0 (the "License"); you may not use
 // this file except in compliance with the License.
-// 
+//
 // You may obtain a copy of the License, together with FAQs at
 // https://www.defold.com/license
-// 
+//
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_collision_object.cpp:7]: I might at some point do a fix for all the licenses with trailing spaces

File: engine/gamesys/src/gamesys/components/comp_collision_object.cpp [hunk 2]
```diff
@@ -1216,16 +1216,16 @@ namespace dmGameSystem
         return GetGroupBitIndex((CollisionWorld*)world, group_hash);
     }
 
-    void RayCast(void* _world, const dmPhysics::RayCastRequest& request, dmPhysics::RayCastResponse& response)
+    void RayCast(void* _world, const dmPhysics::RayCastRequest& request, dmArray<dmPhysics::RayCastResponse>& results)
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_collision_object.cpp:1219]: One design issue was how to allow for an arbitrary number of results.

My first take was to use callbacks, but the scripting interface became a bit too cumbersome.
I then thought of passing a pointer to a pointer, to allow for the function to allocate a buffer, but I instead opted to use our dynamic array class.

I don't feel that any option really were "clean", but this eventually felt cleanest in the end.

File: engine/gamesys/src/gamesys/components/comp_collision_object.h [hunk 3]
```diff
@@ -18,10 +18,7 @@
 #include <gameobject/gameobject.h>
 #include <physics/physics.h>
 
-namespace dmPhysics {
```
Comment (JCash) [inline engine/gamesys/src/gamesys/components/comp_collision_object.h:21]: No need to forward declare if we already include the header!

File: engine/gamesys/src/gamesys/scripts/script_physics.cpp [hunk 4]
```diff
@@ -373,6 +391,13 @@ namespace dmGameSystem
         Vectormath::Aos::Point3 from( *dmScript::CheckVector3(L, 1) );
         Vectormath::Aos::Point3 to( *dmScript::CheckVector3(L, 2) );
 
+        if (Vectormath::Aos::lengthSqr(to - from) <= 0.0f)
+        {
+            dmLogOnceWarning("Ray had 0 length when ray casting, ignoring request.");
+            lua_pushnil(L);
+            return 1;
+        }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_physics.cpp:399]: I moved this check from the c++ libraries, to here, since that's essentially where it's used from.

File: engine/gamesys/src/gamesys/scripts/script_physics.cpp [hunk 5]
```diff
@@ -382,32 +407,61 @@ namespace dmGameSystem
...
+            return_all_results = lua_isnil(L, -1) ? false : lua_toboolean(L, -1);
+            lua_pop(L, 1);
+
+            lua_pop(L, 1);
+
+            list_format = true;
+        }
+
+        dmArray<dmPhysics::RayCastResponse> hits;
+        hits.SetCapacity(32);
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_physics.cpp:426]: Some initial memory

File: engine/gamesys/src/gamesys/scripts/script_physics.cpp [hunk 6]
```diff
@@ -382,32 +407,61 @@ namespace dmGameSystem
...
+            lua_pushvalue(L, 4);
+
+            lua_getfield(L, -1, "all");
+            return_all_results = lua_isnil(L, -1) ? false : lua_toboolean(L, -1);
+            lua_pop(L, 1);
+
+            lua_pop(L, 1);
+
+            list_format = true;
+        }
```
Comment (JCash) [inline engine/gamesys/src/gamesys/scripts/script_physics.cpp:423]: The new api, an options table.

File: engine/particle/src/wscript [hunk 7]
```diff
@@ -17,14 +18,15 @@ def build(bld):
...
+                         protoc_includes = '../proto',
+                         uselib = 'DDF DLIB PLATFORM_SOCKET',
+                         source = 'particle.cpp ../proto/particle/particle_ddf.proto')
 
     bld.install_files('${PREFIX}/include/particle', 'particle.h')
     bld.install_files('${PREFIX}/share/proto', '../proto/particle/particle_ddf.proto')
 
-    bld.add_subdirs('test')
+    if not Options.options.skip_build_tests:
+        bld.add_subdirs('test')
```
Comment (JCash) [inline engine/particle/src/wscript:32]: Unrelated fix: don't build tests when we tell it not to

File: engine/physics/src/physics/physics_3d.cpp [hunk 8]
```diff
@@ -181,6 +197,69 @@ namespace dmPhysics
...
+                }
+                m_hitNormalWorld.push_back(hitNormalWorld);
+                btVector3 hitPointWorld;
+                hitPointWorld.setInterpolate3(m_rayFromWorld,m_rayToWorld,rayResult.m_hitFraction);
+                m_hitPointWorld.push_back(hitPointWorld);
+                m_hitFractions.push_back(rayResult.m_hitFraction);
+                return m_closestHitFraction;
+        }
+    };
+    /// <- END BULLET
```
Comment (JCash) [inline engine/physics/src/physics/physics_3d.cpp:240]: We're currently using an older version of bullet, so I had to copy this class over.

**Review by britzl — DISMISSED** [review]
Looks great!
